reassembly/memory.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2012 Google, Inc. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file in the root of the source
+// tree.
+
+package reassembly
+
+import (
+	"flag"
+	"fmt"
+	//"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"log"
+	"sync"
+	"time"
+)
+
+var memLog = flag.Bool("assembly_memuse_log", defaultDebug, "If true, the github.com/google/gopacket/reassembly library will log information regarding its memory use every once in a while.")
+
+/*
+ * pageCache
+ */
+// pageCache is a concurrency-unsafe store of page objects we use to avoid
+// memory allocation as much as we can.
+type pageCache struct {
+	free         []*page
+	pcSize       int
+	size, used   int
+	pageRequests int64
+	ops          int
+	nextShrink   int
+}
+
+const initialAllocSize = 1024
+
+func newPageCache() *pageCache {
+	pc := &pageCache{
+		free:   make([]*page, 0, initialAllocSize),
+		pcSize: initialAllocSize,
+	}
+	pc.grow()
+	return pc
+}
+
+// grow exponentially increases the size of our page cache as much as necessary.
+func (c *pageCache) grow() {
+	pages := make([]page, c.pcSize)
+	c.size += c.pcSize
+	for i, _ := range pages {
+		c.free = append(c.free, &pages[i])
+	}
+	if *memLog {
+		log.Println("PageCache: created", c.pcSize, "new pages, size:", c.size, "cap:", cap(c.free), "len:", len(c.free))
+	}
+	// control next shrink attempt
+	c.nextShrink = c.pcSize
+	c.ops = 0
+	// prepare for next alloc
+	c.pcSize *= 2
+}
+
+func (c *pageCache) tryShrink() {
+	var min = c.pcSize / 2
+	if min < initialAllocSize {
+		min = initialAllocSize
+	}
+	if len(c.free) <= min {
+		return
+	}
+	for i, _ := range c.free[min:] {
+		c.free[min+i] = nil
+	}
+	c.size -= len(c.free) - min
+	c.free = c.free[:min]
```

Comment:
note:  this free doesn't actually release anything to GC, as far as I know... Go still considers the underlying array and its elements reachable, so it won't GC them.  They're reachable because of http://play.golang.org/p/OpM_EFZria